src/btree/bt_cursor.c

```diff
@@ -530,12 +530,74 @@ __wt_btcur_search_prepared(WT_CURSOR *cursor, WT_UPDATE **updp)
     return (0);
 }
 
+/*
+ * __cursor_reposition_timing_stress --
+ *     Optionally reposition the cursor 10% of times
+ */
+static inline bool
+__cursor_reposition_timing_stress(WT_SESSION_IMPL *session)
+{
+    WT_CONNECTION_IMPL *conn;
+
+    conn = S2C(session);
+
+    if (FLD_ISSET(conn->timing_stress_flags, WT_TIMING_STRESS_EVICT_REPOSITION) &&
+      __wt_random(&session->rnd) % 10 == 0)
+        return (true);
+
+    return (false);
+}
+
+/*
+ * __wt_btcur_evict_reposition --
+ *     Try to evict the page and reposition the cursor on the saved key.
+ */
+int
+__wt_btcur_evict_reposition(WT_CURSOR_BTREE *cbt)
+{
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+
+    cursor = &cbt->iface;
+    session = CUR2S(cbt);
+
+    /*
+     * Try to evict the page and then reposition the cursor back to the page for operations with
+     * snapshot isolation and for pages that require urgent eviction. Snapshot isolation level
+     * maintains a snapshot allowing the cursor to point at the correct value after a reposition
+     * unlike read committed isolation level.
+     */
+    if (session->txn->isolation == WT_ISO_SNAPSHOT && F_ISSET(session->txn, WT_TXN_RUNNING) &&
+      ((__wt_page_evict_soon_check(session, cbt->ref, NULL) &&
+         __wt_hazard_check_count_one(session, cbt->ref)) ||
```

Comment:
If the page is pinned by two cursors, releasing it will only cause us to do a failed eviction and a wasted search again. It's possible that the page is released by both cursors and the eviction can be successful but I think that is very rare.

Reply:
@agorrod, correct me if I am wrong but according to you since the page requires an urgent eviction, maximum effort should be made to try to evict the page, even if it means increase in the latency of the read operation leading to a failure in evicting the page?

Reply:
I believe we should try with the simpler version (not doing the special hazard check), and if that isn't working we can explore mechanisms for making eviction more likely to succeed.

FWIW: I think this hazard check is probably going in the wrong direction most of the time. For a page that needs to be evicted and is in use by multiple threads concurrently, it's most beneficial for threads to back off their use of it. Which allows for eviction to succeed and the page to become less hot. Adding this check in means that individual threads are less likely to back off contentious pages.

Reply:
Then what about yield the thread if we fail the eviction before we do search again so that other threads can have the chance to back off and evict it.